docs/arcadia-users-group/20231128-testing-datasets/lesson.md

```diff
@@ -0,0 +1,84 @@
+# Building test data sets
+
+In our previous lesson, we covered different types of tests.
+In the unit test section, we used five functions to test that our `mean()` function produced the output that we expected it to.
+For each of these functions, we started by defining our test data, `num_list`, a list of numbers for which we wanted to calculate the mean.
+
+These test data were small; each list only had five numbers in it.
+The "correct" answer to the test was clear with relatively little human effort (we could take a few seconds to calculate the mean of each list in our own heads).
+The creation of each test data set was encoded in the test itself.
+
+Below we discuss strategies for creating test data sets to help maintain some of these desirable qualities in real code bases.
+
+## Test data for functions and scripts
+
+### Creating data within the test itself
+
+As seen in our previous lesson and described above, one strategy is to create test data within the test using built-in functions or data structures.
+When possible, this is a great strategy.
+These test data are small and self-contained and the correct answer to the test is easy to figure out.
+Functions like `[]` (list creation), or `range()` in Python and `c()`, `rep()`, `seq()`, `data.frame()`, and `list()` in R can be used to create test data that are relevant to a function.
+Some data-creation functions might be non-deterministic, but setting the seed ensures that the same output are produced each time.
+
+### Testing image outputs
+
+Often times, a function or a script produces an image output.
+Images are trickier to test than some other outputs, but can still be used in unit tests.
+If a function produces the image and an image object is produced in your coding environment, you can sometimes check the metadata of the image to make sure it matches expected values.
+This includes dimensions, color profiles, or file formats.
+These types of tests have the built-in benefit of testing whether the function produces an image successfully on top of confirming attributes about the image.
+
+Even if you check metadata attributes about an image, the image itself might still be different than expected.
+If you want to test the output image more thoroughly, you can compare the output produced by a test to a reference image stored in your test data.
+Both checksum comparisons and pixel comparisons are useful in this scenario.
+
+### Match md5sum of output files
```

Comment:
Consider moving this above 'testing image outputs' -- the progression is (with increasing complexity) ``comparing files directly > comparing md5 > comparing visuals`:

````suggestion
## Comparing file outputs

We can directly compare an output file to a "reference" or "golden" output that is accessible to the test.
Checksums can also be used to compare the results of a function against other types of files.
For example, if you have a function that produces a data frame, you can write that object to a CSV file.
Then, you could compare the output CSV file against a reference CSV file by their checksums.
When using this strategy, it might be necessary to sort your data frame prior to writing a file.

An example of comparing checksums within a test (maybe this should be a challenge?)

```python
import hashlib

with open('expected.txt', 'rb') as expected, '/tmp/propernames.txt', 'rb') as actual:
    assert hashlib.md5(expected.read()).hexdigest() == hashlib.md5(actual.read()).hexdigest()
```

### Testing image outputs
...
````